### PR TITLE
[Adapter] BackendPolicy kind implementation

### PR DESCRIPTION
**Pull request by Amila-Rukshan**
## Purpose
This is to add the BackendPolicy resource for upstream endpoint tls validation. 

 ### Sample configurations
 
 #### Enable tls with a endpoint cert  
 ```
apiVersion: dp.wso2.com/v1alpha1
kind: BackendPolicy
metadata:
  namespace: apk
  name: backendpolicy-sample
spec:
  default:
    protocol: https
    tls: 
      certificateInline: |
        -----BEGIN CERTIFICATE-----
        <cert-data>
        -----END CERTIFICATE-----
  targetRef:
    group: ""
    kind: Service
    namespace: apk
    name: web-hook-service

 ```

#### Enable tls without a endpoint cert (will validate cert using default ca certs)
 ```
apiVersion: dp.wso2.com/v1alpha1
kind: BackendPolicy
metadata:
  namespace: apk
  name: backendpolicy-sample
spec:
  default:
    protocol: https
  targetRef:
    group: ""
    kind: Service
    namespace: apk
    name: web-hook-service
 ```

In this example, changes of this resource or changes of referred config map called `ca-cert-cm-name` are detected.

### Design
https://github.com/wso2/apk/pull/551

### CODE REVIEW

File: adapter/internal/oasparser/model/http_route.go [hunk 1]
```diff
@@ -114,13 +115,16 @@ func (swagger *MgwSwagger) SetInfoHTTPRouteCR(httpRoute *gwapiv1b1.HTTPRoute, au
...
+			}]
 			endPoints = append(endPoints,
-				Endpoint{Host: backendPropertyMapping[types.NamespacedName{
-					Name:      string(backend.Name),
-					Namespace: utils.GetNamespace(backend.Namespace, httpRoute.Namespace),
-				}].ResolvedHostname,
-					URLType: constants.HTTP,
-					Port:    uint32(*backend.Port)})
+				Endpoint{Host: backendProperties.ResolvedHostname,
+					URLType:     oasutils.GetURLType(backendProperties.TLS.Enabled),
```
[the remark this answers is not in the document]
Comment (Amila-Rukshan) [reply adapter/internal/oasparser/model/http_route.go:124]: Yes,  then `backendProperties.TLS.Enabled` is defaulted to `false` and tls will not be enabled

File: adapter/internal/oasparser/utils/open_api_utils.go [hunk 2]
```diff
@@ -97,3 +97,12 @@ func FindAPIDefinitionVersion(jsn []byte) string {
 func FileNameWithoutExtension(path string) string {
 	return strings.TrimSuffix(filepath.Base(path), filepath.Ext(path))
 }
+
+// GetURLType return URLType as https or http
+// depending on whether tls is enabled or not.
+func GetURLType(tlsEnabled bool) string {
```
[the remark this answers is not in the document]
Comment (Amila-Rukshan) [reply adapter/internal/oasparser/utils/open_api_utils.go:103]: moved to a new package called`urlutils`

File: adapter/internal/oasparser/utils/url/url_utils.go [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package urlutils
+
+import "github.com/wso2/apk/adapter/internal/oasparser/constants"
+
+// GetURLType return URLType as https or http
+// depending on whether tls is enabled or not.
+func GetURLType(tlsEnabled bool) string {
+	if tlsEnabled {
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/oasparser/utils/url/url_utils.go:8]: what if we add the protocol from the CRD directly? (HTTP/HTTPS/WS/WSS would be allowed values)

File: adapter/internal/operator/controllers/dp/api_controller.go [hunk 4]
```diff
@@ -315,6 +328,36 @@ func (apiReconciler *APIReconciler) getHostNameForBackend(ctx context.Context, b
...
+	}); err != nil {
+		loggers.LoggerAPKOperator.ErrorC(logging.ErrorDetails{
+			Message:   fmt.Sprintf("Unable to find associated BackendPolicies for service: %s", serviceNamespacedName),
+			Severity:  logging.CRITICAL,
+			ErrorCode: 2611,
+		})
+	}
+	if len(backendPolicyList.Items) > 0 {
+		backendPolicy := backendPolicyList.Items[0]
+		if len(backendPolicyList.Items) > 1 {
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/operator/controllers/dp/api_controller.go:347]: mmm... we can't just ignore other policies. you need to follow policy spec guideline for handling this. // According to https://gateway-api.sigs.k8s.io/geps/gep-713/?h=multiple+targetrefs#conflict-resolution. Refer selectAuthScheme in http_route.go

File: adapter/internal/oasparser/utils/url/url_utils.go (hunk 3 above)
Comment (Amila-Rukshan) [reply adapter/internal/oasparser/utils/url/url_utils.go:8]: Another possibility is to add a property in `Endpoint` struct called `tlsEnabled` and read it to enable TLS?

**Review by AmaliMatharaarachchi — APPROVED** [review]

File: adapter/internal/operator/apis/dp/v1alpha1/api_webhook.go [hunk 5]
```diff
@@ -114,7 +113,10 @@ func (r *API) validateAPIContext() *field.Error {
 			errors.New("unable to list APIs for API context validation"))
 	}
 	for _, api := range apiList.Items {
-		if (utils.NamespacedName(r) != types.NamespacedName{
+		if (types.NamespacedName{
```
[the remark this answers is not in the document]
Comment (Amila-Rukshan) [reply adapter/internal/operator/apis/dp/v1alpha1/api_webhook.go:116]: go a recursive import in test utils due to this import and removed it

**Pull request merged**